File: Chapter 12/Dynamic Scale Beta-t-EGARCH/R/DCStDynamicScaleModeling.R
# Estimate the dynamic-scale Beta-t-EGARCH model on an equity price series.

script_args <- commandArgs(trailingOnly = FALSE)
script_file <- sub("^--file=", "", script_args[grep("^--file=", script_args)])
script_dir <- if (length(script_file)) dirname(normalizePath(script_file)) else getwd()
project_root <- normalizePath(file.path(script_dir, ".."))
data_dir <- file.path(project_root, "Data")
results_dir <- file.path(project_root, "Results")

if (!requireNamespace("readxl", quietly = TRUE)) stop("Install the readxl package before running this script.")
source(file.path(script_dir, "Beta_univ_t_Egarch_Logl.R"))
source(file.path(script_dir, "Beta_univ_t_Egarch_Logl_fact.R"))
source(file.path(script_dir, "Info_Crit.R"))

# This code sets up the dynamic-scale Beta-t-EGARCH model. The active input is the NSX price series.
prices <- as.numeric(readxl::read_excel(file.path(data_dir, "NSX Close---.xlsx"), col_names = FALSE)[[1]])
Y <- diff(log(prices))
t <- length(Y)

# If set to 1, scale dynamics are integrated; otherwise they follow a stationary AR(1) recursion.
I_s <- 0
# If set to 1, standardize the score by its information matrix.
Inf_s <- 1
# If set to 1, add the leverage term kappa_l * sign(-y_t) * (u_t + 1).
lev_s <- 1
Constr_s <- 1
kk_s <- 0.01

# Plot the unconditional distribution and autocorrelation diagnostics.
par(mfrow = c(2, 2))
hist(Y, breaks = 100, probability = TRUE, main = "Unconditional distribution of the data", xlab = "Returns")
curve(dnorm(x, mean(Y), sd(Y)), add = TRUE, col = "darkgreen", lwd = 2)
acf(Y, lag.max = 100, main = "ACF data")
acf(Y^2, lag.max = 100, main = "ACF squared data")
qqnorm(Y, main = "Q-Q plot against Gaussian distribution"); qqline(Y)
par(mfrow = c(1, 1))

# Construct the same starting values and parameter bounds used in the MATLAB script.
if (I_s == 1) {
  if (lev_s == 1) {
    rS <- rep(kk_s, 5); rS[c(1, 2, 4)] <- c(mean(Y), log(sd(Y)), log(8))
    lower <- c(-Inf, -Inf, 0, -Inf, -Inf); upper <- rep(Inf, 5)
  } else {
    rS <- rep(kk_s, 4); rS[c(1, 2, 4)] <- c(mean(Y), log(sd(Y)), log(8))
    lower <- c(-Inf, -Inf, 0, -Inf); upper <- rep(Inf, 4)
  }
} else {
  if (lev_s == 1) {
    rS <- rep(kk_s, 6); rS[c(1, 2, 3, 5)] <- c(mean(Y), log(sd(Y)), .9, log(8))
    lower <- c(-Inf, -Inf, 0, 0, -Inf, -Inf); upper <- c(Inf, Inf, 1, Inf, Inf, Inf)
  } else {
    rS <- rep(kk_s, 5); rS[c(1, 2, 3, 5)] <- c(mean(Y), log(sd(Y)), .9, log(8))
    lower <- c(-Inf, -Inf, 0, 0, -Inf); upper <- c(Inf, Inf, 1, Inf, Inf)
  }
}

# L-BFGS-B is R's bounded optimizer counterpart to constrained MATLAB optimization.
f_s <- function(x) Beta_univ_t_Egarch_Logl(x, Y, Inf_s, I_s)
fit_optim <- optim(rS, f_s, method = "L-BFGS-B", lower = lower, upper = upper, control = list(maxit = 67000))
x_s <- fit_optim$par
dir.create(results_dir, showWarnings = FALSE)
write.csv(data.frame(estimate = x_s), file.path(results_dir, "estPar_R.csv"), row.names = FALSE)

# Obtain filtered scale, residuals, scores, and information criteria.
fitted <- Beta_univ_t_Egarch_Logl_fact(x_s, Y, Inf_s, I_s)
criteria <- Info_Crit(fitted$Logl, x_s, t)
vega_s <- exp(if (I_s == 1) x_s[4] else x_s[5])
mu_s <- x_s[1]

# Plot fitted scale against returns and absolute returns.
par(mfrow = c(2, 1))
plot(Y - mu_s, type = "l", main = "Scale fit: returns", ylab = "Returns / scale")
lines(exp(fitted$lam), col = "red", lwd = 2)
legend("topright", c("Y - mu", "Fitted conditional scale"), col = c("black", "red"), lty = 1)
plot(abs(Y - mu_s), type = "l", main = "Scale fit: absolute returns", ylab = "Absolute returns / scale")
lines(exp(fitted$lam), col = "red", lwd = 2)
par(mfrow = c(1, 1))

# Compare standardized residuals with fitted Student t and Gaussian densities.
hist(fitted$res, breaks = 100, probability = TRUE, main = "Fitted residual distribution", xlab = "Standardized residuals")
grid <- seq(min(fitted$res), max(fitted$res), length.out = 500)
lines(grid, dt(grid, df = vega_s), col = "red", lwd = 2)
lines(grid, dnorm(grid), col = "darkgreen", lwd = 2)
legend("topright", c("Fitted t", "Standard Gaussian"), col = c("red", "darkgreen"), lty = 1)

# Residual, squared-residual, and score ACF diagnostics.
par(mfrow = c(3, 1))
acf(fitted$res, lag.max = 100, main = "ACF residuals")
acf(fitted$res^2, lag.max = 100, main = "ACF squared residuals")
acf(fitted$u, lag.max = 100, main = "ACF fitted scores")
par(mfrow = c(1, 1))

# The analytic Student t CDF is R's counterpart to the numerical MATLAB PIT integration.
PIT_s <- pt(sort(fitted$res), df = vega_s)
plot(PIT_s, type = "l", main = "Ordered PIT DCS model versus uniform", xlab = "Residual rank", ylab = "PIT")
lines(seq(0, 1, length.out = length(PIT_s)), lty = 2)

File: Chapter 12/Dynamic Scale Beta-t-EGARCH/Python/dcst_dynamic_scale_modeling.py
"""Estimate the dynamic-scale Beta-t-EGARCH model on an equity price series."""

from pathlib import Path

import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from scipy.optimize import minimize
from scipy.stats import norm, t as student_t, probplot
from statsmodels.graphics.tsaplots import plot_acf

from beta_univ_t_egarch_logl import beta_univ_t_egarch_logl
from beta_univ_t_egarch_logl_fact import beta_univ_t_egarch_logl_fact
from info_crit import info_crit


PROJECT_ROOT = Path(__file__).resolve().parents[1]
DATA_DIR = PROJECT_ROOT / "Data"
RESULTS_DIR = PROJECT_ROOT / "Results"


def load_log_returns(filename="NSX Close---.xlsx"):
    """Load one price series and convert consecutive prices into log returns."""
    prices = pd.read_excel(DATA_DIR / filename, header=None).to_numpy(dtype=float).reshape(-1)
    return np.log(prices[1:] / prices[:-1])


def initial_parameters(y, integrated, leverage, constrained, kappa_start):
    """Construct the original MATLAB starting values and parameter bounds."""
    if integrated:
        size = 5 if leverage else 4
        start = np.full(size, kappa_start)
        start[0], start[1], start[3] = np.mean(y), np.log(np.std(y, ddof=1)), np.log(8)
        lower, upper = np.full(size, -np.inf), np.full(size, np.inf)
        lower[2] = 0
    else:
        size = 6 if leverage else 5
        start = np.full(size, kappa_start)
        start[0], start[1], start[2], start[4] = np.mean(y), np.log(np.std(y, ddof=1)), 0.9, np.log(8)
        lower, upper = np.full(size, -np.inf), np.full(size, np.inf)
        lower[2], lower[3], upper[2] = 0, 0, 1
    return start, list(zip(lower, upper)) if constrained else None


def plot_exploratory_data(y):
    """Replicate the distribution, ACF, and Q-Q diagnostics from MATLAB."""
    grid = np.linspace(np.min(y), np.max(y), 1000)
    fig, axes = plt.subplots(2, 2, figsize=(12, 8))
    axes[0, 0].hist(y, bins=100, density=True, alpha=0.65, label="Empirical density")
    axes[0, 0].plot(grid, norm.pdf(grid, np.mean(y), np.std(y, ddof=1)), "g-", label="Gaussian")
    axes[0, 0].set_title("Unconditional distribution of the data"); axes[0, 0].legend()
    plot_acf(y, lags=100, ax=axes[0, 1]); axes[0, 1].set_title("ACF data")
    plot_acf(y ** 2, lags=100, ax=axes[1, 0]); axes[1, 0].set_title("ACF squared data")
    probplot(y, dist="norm", plot=axes[1, 1]); axes[1, 1].set_title("Q-Q plot against Gaussian distribution")
    fig.tight_layout()


def run_model(filename="NSX Close---.xlsx", maxiter=67_000, make_plots=True):
    """Run the dynamic-scale model corresponding to DCStDynamicScaleModeling.m."""
    y = load_log_returns(filename)
    integrated_scale, information_scale, leverage_scale = 0, 1, 1
    constrained_scale, kappa_start = 1, 0.01
    if make_plots:
        plot_exploratory_data(y)

    start, bounds = initial_parameters(y, integrated_scale, leverage_scale, constrained_scale, kappa_start)
    objective = lambda par: beta_univ_t_egarch_logl(par, y, information_scale, integrated_scale)
    result = minimize(objective, start, method="L-BFGS-B", bounds=bounds, options={"maxiter": maxiter})
    if not result.success:
        print(f"Optimizer message: {result.message}")

    RESULTS_DIR.mkdir(exist_ok=True)
    pd.DataFrame({"estimate": result.x}).to_excel(RESULTS_DIR / "estPar_Python.xlsx", index=False)
    logl, lam, res, score, beta, fit = beta_univ_t_egarch_logl_fact(result.x, y, information_scale, integrated_scale)
    bic, aic = info_crit(logl, result.x, y.size)
    mu = result.x[0]
    df = np.exp(result.x[3] if integrated_scale else result.x[4])

    if make_plots:
        fig, axes = plt.subplots(2, 1, figsize=(12, 7), sharex=True)
        axes[0].plot(y - mu, label="Y - mu"); axes[0].plot(np.exp(lam), "r", label="Fitted conditional scale")
        axes[0].set_title("Scale fit: returns"); axes[0].legend()
        axes[1].plot(np.abs(y - mu), label="|Y - mu|"); axes[1].plot(np.exp(lam), "r", label="Fitted conditional scale")
        axes[1].set_title("Scale fit: absolute returns"); axes[1].legend(); fig.tight_layout()

        grid = np.linspace(np.min(res), np.max(res), 500)
        plt.figure(figsize=(10, 5)); plt.hist(res, bins=100, density=True, alpha=0.65, label="Empirical residual density")
        plt.plot(grid, student_t.pdf(grid, df), "r-", label="Fitted t")
        plt.plot(grid, norm.pdf(grid), "g-", label="Standard Gaussian")
        plt.title("Empirical fitted residuals versus fitted t distribution"); plt.legend(); plt.tight_layout()

        fig, axes = plt.subplots(3, 1, figsize=(10, 9))
        plot_acf(res, lags=100, ax=axes[0]); axes[0].set_title("ACF residuals")
        plot_acf(res ** 2, lags=100, ax=axes[1]); axes[1].set_title("ACF squared residuals")
        plot_acf(score, lags=100, ax=axes[2]); axes[2].set_title("ACF fitted scores"); fig.tight_layout()

        pit = student_t.cdf(np.sort(res), df)
        plt.figure(figsize=(10, 4)); plt.plot(pit, label="Ordered PIT")
        plt.plot(np.linspace(0, 1, pit.size), "k--", label="Uniform benchmark")
        plt.title("Ordered PIT DCS model versus uniform"); plt.xlabel("Residual rank"); plt.ylabel("PIT"); plt.legend(); plt.tight_layout()

    return {"result": result, "logl": logl, "bic": bic, "aic": aic, "lambda": lam, "residuals": res, "scores": score, "beta": beta, "fit": fit}


if __name__ == "__main__":
    run_model()
    plt.show()
